File: Data Structures & Algorithms/generate-parentheses/submission-2.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
    
        def dfs(l, r, temp):
            
            if l == n and l == r:
                result.append(temp)
            else:
                if l == r and l !=n:
                    dfs(l+1, r, temp+'(')
                if l == n and l > r:
                    dfs(l, r+1, temp+')')
                if l > r and l < n:
                    dfs(l+1, r, temp+'(')
                    dfs(l, r+1, temp+')')
            

        dfs(1, 0, "(")
        return result
            

File: Data Structures & Algorithms/coin-change-ii/submission-2.java
class Solution {
    private final Map<String, Integer> memo = new HashMap<>();

    private int dfs(int amount, int[] coins, int pos) {
        if (amount == 0) return 1;
        if (amount < 0 || pos == coins.length) return 0;

        String key = amount + "-" + pos;
        if (memo.containsKey(key)) return memo.get(key);

        // Option 1: pick current coin
        int include = dfs(amount - coins[pos], coins, pos);

        // Option 2: skip current coin
        int exclude = dfs(amount, coins, pos + 1);

        int total = include + exclude;
        memo.put(key, total);
        return total;
    }

    public int change(int amount, int[] coins) {
        return dfs(amount, coins, 0);
    }
}
